src/pygeos.c

```diff
@@ -0,0 +1,247 @@
+  n2 = GEOSGetNumInteriorRings_r(ctx, geom2);
+  if (n2 == -1) {
+    return 2;
+  }
+  if (n1 != n2) {
+    return 0;
+  }
+
+  /* validate the exterior ring */
+  ring1 = GEOSGetExteriorRing_r(ctx, geom1);
+  if (ring1 == NULL) {
+    return 2;
+  }
+  ring2 = GEOSGetExteriorRing_r(ctx, geom2);
+  if (ring2 == NULL) {
+    return 2;
+  }
+  ret = equals_identical_simple(ctx, ring1, ring2);
+  if (ret != 1) {
+    return ret;
+  }
+  // we don't own the shells, so don't need to destroy them
+
+  /* validate the interior rings */
+  for (i = 0; i < n1; i++) {
+    ring1 = GEOSGetInteriorRingN_r(ctx, geom1, i);
+    if (ring1 == NULL) {
+      return 2;
+    }
+    ring2 = GEOSGetInteriorRingN_r(ctx, geom2, i);
+    if (ring2 == NULL) {
+      return 2;
+    }
+    ret = equals_identical_simple(ctx, ring1, ring2);
+    if (ret != 1) {
+      return ret;
+    }
+  }
+  return 1;
+}
+
+/* Note: forward declaration of the final function that is defined further below,
+so that it can be used in equals_identical_collection just below (due to the
+recursive nature of the implementation). */
+char equals_identical(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                      const GEOSGeometry* geom2);
+
+/* Check the sub-geometries of Multi-part geometries.
+Returns 1 on true, 0 on false, 2 on exception */
+char equals_identical_collection(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                                 const GEOSGeometry* geom2) {
+  int i, n1, n2;
+  const GEOSGeometry *sub_geom1 = NULL;
+  const GEOSGeometry *sub_geom2 = NULL;
+  char ret;
+
+  n1 = GEOSGetNumGeometries_r(ctx, geom1);
+  if (n1 == -1) {
+    return 2;
+  }
+  n2 = GEOSGetNumGeometries_r(ctx, geom2);
+  if (n2 == -1) {
+    return 2;
+  }
+  if (n1 != n2) {
+    return 0;
+  }
+
+  /* validate the sub geometries */
+  for (i = 0; i < n1; i++) {
+    sub_geom1 = GEOSGetGeometryN_r(ctx, geom1, i);
+    if (sub_geom1 == NULL) {
+      return 2;
+    }
+    sub_geom2 = GEOSGetGeometryN_r(ctx, geom2, i);
+    if (sub_geom2 == NULL) {
+      return 2;
+    }
+    ret = equals_identical(ctx, sub_geom1, sub_geom2);
+    if (ret != 1) {
+      return ret;
+    }
+  }
+  return 1;
+}
+
+/* Check geometries for being identical.
+Returns 1 on true, 0 on false, 2 on exception */
+char equals_identical(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                      const GEOSGeometry* geom2) {
+  int type1, type2;
+
+  type1 = GEOSGeomTypeId_r(ctx, geom1);
+  if (type1 == -1) {
+    return 2;
+  }
+  type2 = GEOSGeomTypeId_r(ctx, geom2);
+  if (type2 == -1) {
+    return 2;
+  }
+  if (type1 != type2) {
+    return 0;
+  }
+
+  if ((type1 == 0) || (type1 == 1) || (type1 == 2)) {
+    return equals_identical_simple(ctx, geom1, geom2);
+  } else if (type1 == 3) {
+    return equals_identical_polygon(ctx, geom1, geom2);
+  } else if ((type1 >= 4) && (type1 <= 7)) {
+    return equals_identical_collection(ctx, geom1, geom2);
+  } else {
+    return 2;
+  }
```

Comment:
Is it more readable using enums? And what about a switch?
```suggestion
  switch (type1) {
    case GEOS_POINT:
    case GEOS_LINESTRING:
    case GEOS_LINEARRING:
      return equals_identical_simple(ctx, geom1, geom2);
    case GEOS_POLYGON:
      return equals_identical_polygon(ctx, geom1, geom2);
    case GEOS_MULTIPOINT:
    case GEOS_MULTILINESTRING:
    case GEOS_MULTIPOLYGON:
    case GEOS_GEOMETRYCOLLECTION:
      return equals_identical_collection(ctx, geom1, geom2);
    default:
      return 2;
  }
```

Reply:
Yes, that's indeed nicer